gapic-common/lib/gapic/rest/server_stream.rb

```diff
@@ -0,0 +1,104 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "json"
+
+module Gapic
+  module Rest
+    ##
+    # A class to provide the Enumerable interface to the response of a REST server-streaming dmethod.
+    #
+    # ServerStream provides the enumerations over the individual response messages within the stream.
+    #
+    # @example normal iteration over resources.
+    #   server_stream.each { |response| puts response }
+    #
+    class ServerStream
+      include Enumerable
+
+      # @param message_klass [Class]
+      # @param json_enumerator [Enumerator<String>]
+      def initialize message_klass, json_enumerator
+        @json_enumerator = json_enumerator
+        @_obj = ""
+        @message_klass = message_klass
+        @ready_objs = [] # List of strings
+      end
+
+      ##
+      # Iterate over JSON objects in the streamed response.
+      #
+      # @yield [Object] Gives one complete Message object.
+      #
+      # @return [Enumerator] if no block is provided
+      #
+      def each
+        return enum_for :each unless block_given?
+
+        loop do
+          while @ready_objs.length.zero?
+            begin
+              chunk = @json_enumerator.next
+              next unless chunk
+              _next_json! chunk
+            rescue StopIteration
+              dangling_content = @_obj.strip
+              error_expl = "Dangling conent left after iterating through the stream. " \
+                           "This means that not all content was received or parsed correctly. " \
+                           "It is likely a result of server or network error."
+              error_text = "#{error_expl}\n Content left unparsed: #{dangling_content}"
+
+              raise Gapic::Common::Error, error_text unless dangling_content.empty?
+              return
+            end
+          end
+          yield @message_klass.decode_json @ready_objs.shift, ignore_unknown_fields: true
+        end
+      end
+
+      private
+
+      ##
+      # Builds the next JSON object of the server stream from chunk.
+      #
+      # @param chunk [String] Contains (partial) JSON object
+      #
+      def _next_json! chunk
+        chunk.chars.each do |char|
+          @_obj += char
+          # Invariant: @_obj is always either a part of a single JSON object or the entire JSON object.
+          # Hence, it's safe to strip whitespace, commans and array brackets. These characters
+          # are only added before @_obj is a complete JSON object and essentially can be flushed.
+          @_obj = @_obj.lstrip # strip whitespace.
+          # Eat array delimiter characters.
+          if @_obj[0] == "[" || @_obj[0] == "," || @_obj[0] == "]"
+            @_obj = @_obj[1..]
+          end
```

Comment:
I think we can optimize this a lot better. Currently we're doing a lot of calls, including creating multiple new objects and copying a lot of string data, on every character added. This makes for a very slow inner loop. Can we instead:

* Replace the logic in lines 83-87 with a single call like `@_obj.sub!(/^([\s\[\],)+/, "")`
* Lift that call out of the loop. It should only need to happen once, to flush out those characters before we start reading new ones.
* Then split the each loop into two phases: the first that reads characters in the skip list (space, brackets, and commas) and throws them away, and then the second that starts with the first real character and does what you're doing below.

Reply:
Thanks for the suggestion. I don't understand the last part. Do you mean split each loop into two each loops ?

Reply:
Something else you might consider is: `chunk.scan(/\{|\}|[^{}]+/) do |chrs| ...`. That would reduce the number of iterations significantly, by combining runs of characters that aren't braces.

Reply:
Thanks so much for thorough explanation. I added an optimization you suggested to skip unneccesary string concatenations. I used `char.strip` to check for whitespaces - do you think it's slow?

Reply:
Following up: I suggest `next if @obj.empty? && char != "{"` to replace lines 85-87.

Reply:
Incredible. Thanks.